currencies/rate_converter.go

```diff
@@ -0,0 +1,205 @@
+		fetchingInterval,
+		retryFetchingInterval,
+		nil, // no notifier channel specified, won't send any notifications
+	)
+}
+
+// NewRateConverterWithNotifier returns a new RateConverter
+// it allow to pass an update chan in which the number of ticks will be passed after each tick
+// allowing clients to listen on updates
+// Do not use this method
+func NewRateConverterWithNotifier(
+	httpClient httpClient,
+	syncSourceURL string,
+	fetchingInterval time.Duration,
+	retryFetchingInterval time.Duration,
+	updateNotifier chan<- int,
+) *RateConverter {
+	rc := &RateConverter{
+		httpClient:            httpClient,
+		updates:               make(chan *Rates),
+		done:                  make(chan bool),
+		updateNotifier:        updateNotifier,
+		fetchingInterval:      fetchingInterval,
+		retryFetchingInterval: retryFetchingInterval,
+		ticker:                nil,
+		syncSourceURL:         syncSourceURL,
+		rates:                 atomic.Value{},
+		lastUpdated:           atomic.Value{},
+	}
+
+	// In case host do not want to support currency lookup
+	// we just stop here and do nothing
+	if rc.fetchingInterval == time.Duration(0) {
+		return rc
+	}
+
+	rc.Update()                // Make sure to populate data before to return the converter
+	rc.startPeriodicFetching() // Start periodic ticking
+
+	return rc
+}
+
+// fetch allows to retrieve the currencies rates from the syncSourceURL provided
+func (rc *RateConverter) fetch() (*Rates, error) {
+	request, err := http.NewRequest("GET", rc.syncSourceURL, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+	request = request.WithContext(ctx)
+
+	response, err := rc.httpClient.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	defer response.Body.Close()
+
+	bytesJSON, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	updatedRates := &Rates{}
+	err = json.Unmarshal(bytesJSON, &updatedRates)
+	if err != nil {
+		return nil, err
+	}
+
+	return updatedRates, err
+}
+
+// updateRates allows to update the internal currencies rates map with the one provided
+func (rc *RateConverter) updateRates(updatedRates *Rates) {
+	rc.rates.Store(updatedRates)
+	rc.lastUpdated.Store(time.Now())
+}
+
+// Update updates the internal currencies rates from remote sources
+func (rc *RateConverter) Update() (*Rates, error) {
+	rates, err := rc.fetch()
+	if err == nil {
+		rc.updateRates(rates)
+	}
+
+	return rates, err
+}
+
+// startPeriodicFetching starts the periodic fetching at the given interval
+// triggers a first fetch when called before the first tick happen in order to initialize currencies rates map
+// returns a chan in which the number of data updates everytime a new update was done
+func (rc *RateConverter) startPeriodicFetching() error {
+
+	rc.ticker = time.NewTicker(rc.fetchingInterval)
+	updatesTicksCount := 0
+
+	go func() {
+		for {
+			select {
+			case <-rc.ticker.C:
```

Comment:
Would it make sense to call `rc.Update()` here rather than spawning off two go routines to handle one update? Do these even need to be handled in go routines? All that it would allow is multiple updates to happen in parallel if necessary, and I don't see a use case for that.

I would suggest just calling `rc.Update()` on a tick, so you can have all your update code in one spot. That will allow the initial `Update()` call benefit from any retry logic implemented in the future. If a new update tick comes along while one update is still processing, it doesn't seem useful to spawn a second update attempt. The context created within the Update code will handle the case where one request hangs, so it will eventually be retried.

Reply:
Indeed, it would be way simpler; I updated it with what you suggested, let me know if I did it right.

Reply:
Looks good to me. :)